src/test/java/spoon/test/generics/GenericsTest.java

```diff
@@ -695,36 +725,39 @@ public void testCtSuperTypeHierarchy() throws Exception {
+		CtSuperTypeHierarchy typingContextOfDisgust = new CtSuperTypeHierarchy(typeReferenceOfDisgust);
 		
-		//adapt X to scope of CelebrationLunch<Integer,Long,Double>.WeddingLunch<Mole>
-		CtTypeReference<?> adapted = sthOftrWeddingLunch_Mole.adaptType(ctWeddingLunch_X);
-		assertEquals("spoon.test.generics.testclasses.Mole", adapted.getQualifiedName());
+		// in disgust, X of WeddingLunch is bound to "Model"
+		assertEquals("spoon.test.generics.testclasses.Mole", typingContextOfDisgust.adaptType(ctWeddingLunch_X).getQualifiedName());
 		//adapt A to scope of CelebrationLunch<Integer,Long,Double>.WeddingLunch<Mole>
-		adapted = sthOftrWeddingLunch_Mole.adaptType(ctClassLunch_A);
-		assertEquals("spoon.test.generics.testclasses.Mole", adapted.getQualifiedName());
-		//adapt A to scope of enclosing class of CelebrationLunch<Integer,Long,Double>.WeddingLunch<Mole>, which is CelebrationLunch<Integer,Long,Double>
-		adapted = sthOftrWeddingLunch_Mole.getEnclosingGenericTypeAdapter().adaptType(ctClassLunch_A);
-		assertEquals("java.lang.Double", adapted.getQualifiedName());
-		
+
+		// in disgust, the A of Lunch is bound to "Mole"
+		assertEquals("spoon.test.generics.testclasses.Mole", typingContextOfDisgust.adaptType(ctClassLunch_A).getQualifiedName());
+
+		// I don't understand the goal and utility of this one
+		assertEquals("java.lang.Double", typingContextOfDisgust.getEnclosingGenericTypeAdapter().adaptType(ctClassLunch_A).getQualifiedName());
+
+
+		// now we resolve those types, but in the context of the declaration, where no concrete types exist
 		//are these two types same in scope of CelebrationLunch<K,L,M>.WddingLunch<X> class itself
-		CtSuperTypeHierarchy sthOftWeddingLunch_X = new CtSuperTypeHierarchy(trWeddingLunch_Mole.getDeclaration());
+		CtSuperTypeHierarchy sthOftWeddingLunch_X = new CtSuperTypeHierarchy(typeReferenceOfDisgust.getDeclaration());
 		
-		//adapt X to scope of type CelebrationLunch<K,L,M>.WddingLunch<X>
-		adapted = sthOftWeddingLunch_X.adaptType(ctWeddingLunch_X);
-		assertEquals("X", adapted.getQualifiedName());
-		//adapt A to scope of type CelebrationLunch<K,L,M>.WddingLunch<X>
-		adapted = sthOftWeddingLunch_X.adaptType(ctClassLunch_A);
-		assertEquals("X", adapted.getQualifiedName());
+		// in WeddingLunch "X" is still "X"
+		assertEquals("X", sthOftWeddingLunch_X.adaptType(ctWeddingLunch_X).getQualifiedName());
+
+		// in WeddingLunch the "A" from Lunch of is called "X"
+		assertEquals("X", sthOftWeddingLunch_X.adaptType(ctClassLunch_A).getQualifiedName());
+
+		// ?????
 		//adapt A to scope of enclosing class of CelebrationLunch<K,L,M>.WddingLunch<X>, which is CelebrationLunch<K,L,M>
-		adapted = sthOftWeddingLunch_X.getEnclosingGenericTypeAdapter().adaptType(ctClassLunch_A);
-		assertEquals("M", adapted.getQualifiedName());
+		assertEquals("M", sthOftWeddingLunch_X.getEnclosingGenericTypeAdapter().adaptType(ctClassLunch_A).getQualifiedName());
```

Comment:
Same like above. In this simplified model
```
class Lunch<A,B> {}
class CelebrationLunch<K,L,M> extends Lunch<M,K> {}
```
it shows that type parameter `A` of `Lunch` is adapted to `M` of `CelebrationLunch`.